contk/dataloader/dataloader.py

```diff
@@ -26,3 +30,255 @@ def load_class(cls, class_name):
+		if batch_size is None and self.batch_size[key] is None:
+			raise ValueError("You need batch_size to intialize.")
+		if shuffle:
+			random.shuffle(self.index[key])
+
+		self.batch_id[key] = 0
+		if batch_size is not None:
+			self.batch_size[key] = batch_size
+		print("%s set restart, %d batches and %d left" % (key, \
+				len(self.index[key]) // self.batch_size[key], \
+				len(self.index[key]) % self.batch_size[key]))
+
+	def get_batch(self, key, index):
+		'''Get a batch of specified `index`.
+
+				Arguments:
+						key (str): must be contained in `key_name`
+						index (list): a list of specified index
+
+				Returns:
+						A dict at least contains ``sentence``, ``sentence_length``. See the example belows.
+		'''
+		raise NotImplementedError( \
+			"This function should be implemented by subclasses.")
+
+	def get_next_batch(self, key, ignore_left_samples=False):
+		'''Get next batch.
+
+		Arguments:
+				key (str): must be contained in `key_name`
+				ignore_left_samples (bool): Ignore the last batch, whose sample num
+						is not equal to `batch_size`. Default: `False`
+
+		Returns:
+				A dict like :func:`get_batch`, or None if the epoch is end.
+		'''
+		if key not in self.key_name:
+			raise ValueError("No set named %s." % key)
+		if self.batch_size[key] is None:
+			raise RuntimeError( \
+				"Please run restart before calling this function.")
+		batch_id = self.batch_id[key]
+		start, end = batch_id * \
+			self.batch_size[key], (batch_id + 1) * self.batch_size[key]
+		if start >= len(self.index[key]):
+			return None
+		if ignore_left_samples and end > len(self.index[key]):
+			return None
+		index = self.index[key][start:end]
+		res = self.get_batch(key, index)
+		self.batch_id[key] += 1
+		return res
+
+	def sen_to_index(self, sen):
+		'''Convert a sentences from string to index representation.
+
+		Arguments:
+			sen (list): a list of str, representing each token of the sentences.
+
+		Examples:
+			vocab_list = ["<pad>", "<unk>", "<go>", "<eos>", "I", "have",\
+							"been", "to", "Sichuan"]
+			>>> dataloader.sen_to_index(
+			...		["<go>", "I", "have", "been", "to", "Sichuan", "<eos>"])
+			>>> [2, 4, 5, 6, 7 ,8 ,3]
+
+		'''
+		return list(map(lambda word: self.word2id.get(word, self.unk_id), sen))
+
+	def trim_index(self, index):
+		'''Trim index. There will be two steps:
+			* If there is an `<eos>` in sentences, \
+				find first `<eos>` and abondon words after it (included the `<eos>`).
+			* ignore `<pad>` s at the end of the sentence.
+
+		Arguments:
+			index (list): a list of int
+
+		Examples:
+			vocab_list = ["<pad>", "<unk>", "<go>", "<eos>", "I", "have",\
+							"been", "to", "Sichuan"]
+			>>> dataloader.trim_index(
+			...		[2, 4, 5, 6, 7, 8, 0, 0, 3, 4, 3, 0]) # <go> I have been to Sichuan <pad> <pad> <eos> I <eos> <pad>
+			>>> [2, 4, 5, 6, 7, 8] # <go> I have been to Sichuan
+		'''
+
+		index = trim_before_target(list(index), self.eos_id if not self.multi_turn else self.eot_id)
```

Comment:
You must explain where ``multi_turn`` come from. Or it will be difficult for others to understand how to realize a subclass.
I prefer making it a parameter of ``__init__``

Reply:
Also, using ``multi_turn`` here is lack of scalability. Imagine we have another dataloader, the end token is not `eos_id` or `eot_id`.

Reply:
Yes, I think it is really a problem. So, maybe should I change the `trim_index` function?

Or change the parameters of `__init__`?

Reply:
Got it.